flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -1093,6 +1098,23 @@ public boolean restoreLatestCheckpointedState(
 				}
 			}
 
+			CompletedCheckpoint candidate = null;
+			if (isPerferCheckpointForRecovery && completedCheckpointStore.getAllCheckpoints().size() > 1) {
```

Comment:
It feels like this part basically reimplements the getLatestCheckpoint() logic of the completed checkpoint store. Maybe instead of having this hear the getLatestCheckpoint( ) could take the preferCheckpoints boolean parameter for better encapsulation

Reply:
Agree.